### DISCUSSION TITLE
Deploying Next.js to IIS with a custom server

### DISCUSSION THREAD

**Discussion by Kudratov** [vercel/next.js#13146] [Help] [answered]
Hi everyone
Is there any way to deploy Next js app to IIS, nearly there are no docs those are related to the deployment if you have any idea or docs that will help to solve the issue, please share.
Any answer will be appreciated.

**Comment by timneutkens** [comment, marked as answer]
`next build` and then `next start` should be enough, it's a Node.js application so it's not any different from hosting a Node.js server on IIS

**Reply by emmsdan** [reply]
Hi @timneutkens 
Am using IISnode, and in there I need to point to a specific file ie. server.js
```xml
<handlers>
      <add name="iisnode" path="dist/server.js" verb="*" modules="iisnode" />
</handlers>
```
after building the next app `next build` where can I point `node` to directly. `node next/server.js`

**Reply by timneutkens** [reply]
`node ./node_modules/bin/next start`

**Reply by rpateld** [reply]
should be something ".next/server/init-server.js.js" but it's not working. Please confirm.

**Comment by mcmillhj-wta** [comment]
@Kudratov how did you handle the port? That is what I am currently stuck on

**Comment by nlimpag07** [comment]
Hi, I think we need clarification on this. I also encountered similar and I've seen that the webpack-hmr still runs in production.

Server: IIS via iisnode

In package.json file:
```
"scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "node server.js"
  },
```
  
  In server.js:
```
  // server.js
const { createServer } = require('http')
const { parse } = require('url')
const next = require('next')

const dev = process.env.NODE_ENV !== 'production'
const port = process.env.PORT || 3000;
const app = next({ dev })
const handle = app.getRequestHandler()

app.prepare().then(() => {
  createServer((req, res) => {
    // Be sure to pass `true` as the second argument to `url.parse`.
    // This tells it to parse the query portion of the URL.
    const parsedUrl = parse(req.url, true)
    const { pathname, query } = parsedUrl

    if (pathname === '/a') {
      app.render(req, res, '/a', query)
    } else if (pathname === '/b') {
      app.render(req, res, '/b', query)
    } else {
      handle(req, res, parsedUrl)
    }
  }).listen(port, (err) => {
    if (err) throw err
    console.log(`> Ready on http://localhost:${port}`)
  })
})
```


In web.config:
```
<configuration>
  <system.webServer>    
    <rewrite>
      <rules>
        <rule name="myapp">
          <match url="/*" />
          <action type="Rewrite" url="server.js" />
        </rule>
      </rules>
    </rewrite>

    <iisnode node_env="production" nodeProcessCommandLine="&quot;E:\GIT\node.exe&quot;" interceptor="&quot;%programfiles%\iisnode\interceptor.js&quot;" />

  </system.webServer>
    <location path="" overrideMode="Deny">
        <system.webServer>
    <handlers>
      <add name="iisnode" path="server.js" verb="*" modules="iisnode" />
    </handlers>
        </system.webServer>
    </location>
</configuration>
```

This works but there are performance issues as webpack-hmr still runs in production.

But if we change the package,json  "start": "node server.js" into  "start": "next start" then build and run in production, the webpack-hmr won't run (the expected).

Is there any way to run  "next start" from web.config without going through server.js?

**Reply by Flija0** [reply]
But what if i am not using localhost ? and instead a custom domain

**Comment by ShawnCentra** [comment]
Also curious about deployment on IIS with next.js 

I get that its a node.js application but there seems to be a lack of documentation around the setup, seems overtly complicated

**Reply by ShawnCentra** [reply]
Update: next.js npm build -> npm start works with IISnode with no config server ( no custom server )

**Reply by hagenderouen** [reply]
@ShawnCentra Please share your webconfig. Thanks

**Reply by ShawnCentra** [reply]
`<configuration>
    <system.webServer>
        <rewrite>
            <rules>					
                <rule name="ReverseProxyInboundRule1" stopProcessing="true">
                    <match url="(.*)" />
                    <action type="Rewrite" url="http://localhost:3000/{R:1}" />
                </rule>
            </rules>
        </rewrite>
    </system.webServer>
</configuration>`

Set up reverse proxy so that the node server handles the routing etc. on IIS

**Reply by ShawnCentra** [reply]
![image](https://user-images.githubusercontent.com/78757056/150175976-0e2a7403-e8b2-47c6-8c57-344a5f9b153e.png)

the http redirect one isn't necessary for it to get running just the one below it 

I followed some reverse proxy set up guide with IIS online, dont remember the link I'll try to find it later.

**Reply by AnweshGangula** [reply]
Found following resources that can be helpful: 
- https://alex.domenici.net/archive/deploying-a-node-js-application-on-windows-iis-using-a-reverse-proxy
- https://stackoverflow.com/a/63098601/6908282
- https://stackoverflow.com/q/54965069/6908282

**Comment by bishrant** [comment]
For someone coming here and being frustrated with the approved answer. Here are the config that worked for us
-You need to install IISNode and URLRewrite module.
1. Set the permissions on whole folder to be read/write to IIS_IUSRS
2. run `npm run build`
3. Create app.js as your main point of entry
```js
const { createServer } = require('http')
const { parse } = require('url')
const next = require('next')

const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev })
const handle = app.getRequestHandler()

app.prepare().then(() => {
  createServer((req, res) => {
    const parsedUrl = parse(req.url, true)
      handle(req, res, parsedUrl)
  }).listen(process.env.PORT || 3000, (err) => {
    if (err) throw err
    console.log('> Ready on http://localhost')
  })
})
```

4. Create `web.config` on your root with following content
```xml
<configuration>
  <system.webServer>
    <handlers>
      <add name="iisnode" path="app.js" verb="*" modules="iisnode" />
    </handlers>
    <iisnode loggingEnabled="true" debuggingEnabled="false"/>
    <rewrite>
      <rules>
        <rule name="app" patternSyntax="Wildcard">
          <match url="*"  />
          <action type="Rewrite" url="app.js" />
        </rule>
      </rules>
    </rewrite>
  </system.webServer>
</configuration>
```

_**Make sure you copy the `node_modules` and `.next` folder. You might be able to strip more folders but we haven't got that far.**_

If you run into any issues, usually deleting .next and node_modules folder followed by IIS restart and `npm install` and `npm run build` fixes things

**Reply by ShawnCentra** [reply]
If you have a custom server yea this is the way to go.

**Reply by hagenderouen** [reply]
I'm missing the solution without a customer server

**Reply by ShawnCentra** [reply]
@hagenderouen  reverse proxy node application set up 
[Reverse Proxy](https://dev.to/petereysermans/hosting-a-node-js-application-on-windows-with-iis-as-reverse-proxy-397b)

**Reply by webdeveloper2794** [reply]
that worked for me thanks

**Comment by q7314568** [comment]
How can i host next on IISNode with SSL?

**Reply by ShawnCentra** [reply]
reverse proxy guide from above, follow my config bellow and setup HTTPS through IIS bindings

**Comment by zafergurel** [comment]
If you want to deploy the html export of your next.js site, I've written my experience [here](https://dev.to/zaff/how-to-properly-deploy-a-static-html-site-built-by-nextjs-on-iis-to-avoid-404s-1pce).
By deploying the html export only, you don't need to deal with node.js and other stuff.

**Reply by q7314568** [reply]
thx for reply,have u tried deploy with serverside?

**Reply by zafergurel** [reply]
Yes, I have. That's why I wrote that blog post to share the experience.

**Comment by AbdelazizEid** [comment]
[This guide](https://dev.to/petereysermans/hosting-a-node-js-application-on-windows-with-iis-as-reverse-proxy-397b) helps me to run next start on IIS without needing server.js  or next export

**Comment by OZZlE** [comment]
`const port = process.env.PORT || process.env.port;` is key to making it work with iisnode, I guess it decides itself which port it wants to use, if you try to define it yourself it seems to always fail..

**Reply by cmolina-rugertek** [reply]
That worked for me! I was stuck for two days. Than you so much!

**Comment by ShawnCentra** [comment]
Hey guys. 

You need pm2 
IISnode installed on IIS 

commands I run to build and deploy (i use yarn, but you can use npm with npm run ~command~) 
yarn build:production
yarn pm2
```
//package.json
{
  "name": "website",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "build:production": "cp .env.prod .env.local && next build --debug",
    "build:staging": "cp .env.staging .env.local && next build --debug",
    "dev:staging": "cp .env.staging .env.local && next dev",
    "postbuild": "npx next-sitemap --config custom-next-sitemap.js",
    "start": "next start",
    "devServer": "node server.js",
    "pm2": "pm2 start pm2.json"
  },
```

```
//web.config file at root directory of project (next to package.json)
<configuration>
    <system.webServer>
        <rewrite>
            <rules>
                <clear />
                <rule name="ReverseProxyInboundRule1" stopProcessing="true">
                    <match url="(.*)" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="false" />
                    <action type="Rewrite" url="http://localhost:3033/{R:1}" />
                </rule>
            </rules>
        </rewrite>
    </system.webServer>
</configuration> 

```


```
//pm2.json (root dir)
{
	"apps": [
		{
			"name": "website",
			"script": "node_modules/next/dist/bin/next",
			"log_date_format" : "YYYY-MM-DD HH:mm Z",
			"args": "start --port 3033",
			"cwd": "./"
		}
	]
} 



```

**Reply by ShawnCentra** [reply]
point IIS site to the directory 
set up the port 80 and port 443 for https and add the domain name for the site (otherwise it'll deploy on localhost)

**Reply by ShawnCentra** [reply]
you can see if your site is started on pm2 using `pm2 list`

**Comment by Smirra** [comment]
If anyone wants the default standalone server to work on newer versions of nextjs (13+) on iisnode, you need remove the `parseInt` function when setting the port in the `server.js`.
So
`const currentPort = parseInt(process.env.PORT, 10) || 3000`
Becomes
`const currentPort = process.env.PORT || 3000`

This totally fixed it for me.

**Comment by UTSProjects** [comment]
I am facing an issue: I have successfully deployed my nextjs project  to IIS.  It is a subdomain project to let's say www.xyz.com and my project is a sub project called Jam. Now the issue is that , when i am accessing the project through Url: https://xyz.com/Jam then its working fine, but when I do  https://www.xyz.com/Jam then it shows an error : Web server is returning an unknown error error code: 520.

Below is my web.config file : 
<configuration>
  <system.webServer>    
    <rewrite>
      <rules>
        <rule name="Rewrite to server.js" enabled="true">
          <match url=".*" ignoreCase="true" />
          <action type="Rewrite" url="server.js" />
        </rule>
      </rules>
    </rewrite>

    <iisnode node_env="production" nodeProcessCommandLine="&quot;C:\Program Files\nodejs\node.exe&quot;" interceptor="&quot;%programfiles%\iisnode\interceptor.js&quot;" />
        <modules>
            <add name="iisnode" />
        </modules>

  </system.webServer>
    <location path="" overrideMode="Deny">
        <system.webServer>
    <handlers>
      <add name="iisnode" path="server.js" verb="*" modules="iisnode" />
    </handlers>
        </system.webServer>
    </location>
</configuration>


And this is my next.config.mjs file : 
const { basePath } = require("./Jams.js");

module.exports = {
  output: "standalone",
  basePath: `/${basePath}`, // Use basePath from Jams.js
  assetPrefix: `/${basePath}`, // Use basePath as prefix for assets
  async rewrites() {
    return [
      {
        source: '/:path*',
        destination: '/:path*',
      },
      {
        source: `/${basePath}/:path*`, // Correct use of backticks for string interpolation
        destination: `/${basePath}/:path*`,
      },
    ];
  },
  webpack: (config, { isServer }) => {
    // Add raw-loader for txt files and node files
    config.module.rules.push(
      {
        test: /\.txt$/, // Test for .txt files
        use: "raw-loader",
      },
      {
        test: /\.node$/, // Test for .node files
        use: "raw-loader",
      }
    );

    return config;
  },
};

I cant understand where the problem is, Please Help.
